### DISCUSSION TITLE
existing column value is not getting copied to newly added column

### DISCUSSION THREAD

**Discussion by rajatgng** [sqlalchemy/sqlalchemy#7572] [Usage Questions] [answered]
### Describe the bug

existing column value is not getting copied to a newly added column in migration

### To Reproduce

```python
"""add activated_at and activated_by_id in campaigns table

Revision ID: a20acc3a7fc6
Revises: ba0039600ae0
Create Date: 2022-01-14 15:33:01.685904

"""
from alembic import op
import sqlalchemy as sa
from sqlalchemy import orm
import app

# revision identifiers, used by Alembic.
revision = "a20acc3a7fc6"
down_revision = "ba0039600ae0"
branch_labels = None
depends_on = None

#Note: In "campaigns" table, "created_at" column is already there with value populated in it.

def upgrade():
    op.add_column(
        "campaigns",
        sa.Column(
            "activated_at",
            sa.DateTime(timezone=True),
            nullable=True,
        ),
    )
    
    # Populate `activated_at` field on existing records based on `active` field value
    bind = op.get_bind()
    session = orm.Session(bind=bind)

    for campaign in session.query(app.models.Campaign).all():
        if campaign.active:
            campaign.activated_at = campaign.updated_at

    session.commit()


def downgrade():
    op.drop_column("campaigns", "activated_at")
```

### Error

```
When checking the database activated_at is already present but it's value is null instead of value that is present in updated_at
``` 

### Versions

- OS: Ubuntu 20
- Python: 3.8.10
- SQLAlchemy: 1.3
-alembic==1.5.5
- Database: postgres
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): psycopg

**Comment by zzzeek** [comment]
hey there -

SQLAlchemy ORM models don't automatically update themselves to include mappings to newly added columns.  if you wanted to use the pattern you have here (which I would advise against, there's a much better way see below), you'd need to update the mapping first, something like this, assuming you are using declarative mappings with declarative_base():

```py
def upgrade():
    op.add_column(
        "campaigns",
        sa.Column(
            "activated_at",
            sa.DateTime(timezone=True),
            nullable=True,
        ),
    )
    
    # Populate `activated_at` field on existing records based on `active` field value
    bind = op.get_bind()
    session = orm.Session(bind=bind)

    # this will only work if you are using a declarative Base class from the
    # declarative_base() function
    app.models.Campaign.activated_at = Column("activated_at", DateTime(timezone=True))

    for campaign in session.query(app.models.Campaign).all():
        if campaign.active:
            campaign.activated_at = campaign.updated_at

    session.commit()
```

However, the above approach is not a best practice for a database migration.   when doing migrations you want to avoid round-trips from the database to the client as much as possible, and in SQL, setting a column to be the same value of another one on a per-row basis is very simple using SQL UPDATE.  so the better migration would be this:

```py
from sqlalchemy import table, column

def upgrade():
    op.add_column(
        "campaigns",
        sa.Column(
            "activated_at",
            sa.DateTime(timezone=True),
            nullable=True,
        ),
    )
    table = table("campaigns", column("activated_at"), column("updated_at"))
    op.get_bind().execute(table.update().values(activated_at = table.c.updated_at))
```

that will run a single UPDATE that doesn't incur any database round trips, and avoids the complexity of using ORM models inside of migrations which is generally inconvenient and usually not necessary.

**Reply by rajatgng** [reply]
Thanks, @zzzeek, this work as well. By the way, the below approach which I found easier it is just a single line, Is this approach is good?

`
op.execute('UPDATE campaigns SET activated_at = updated_at;')
`

**Comment by rajatgng** [comment, marked as answer]
`op.execute('UPDATE campaigns SET activated_at = updated_at;')`

Adding above line in upgrade function after the add_column work well for me.

**Reply by zzzeek** [reply]
yes you can do it that way too.  for forwads-compat, use the text() construct, also omit the semicolon some drivers dont like it:

   op.execute(text('UPDATE campaigns SET activated_at = updated_at'))

**Reply by rajatgng** [reply]
Thanks for the detailed explanation, really appreciate it @zzzeek.
